### ISSUE TITLE
[Enhancement] Add sync endpoint for Kobo device

### ISSUE THREAD

**Issue by shavitmichael** [janeczku/calibre-web#1077]
It's trivial to change the endpoint that the Kobo device talks to for its cloud syncing protocol.
The Sync API allows the Kobo device to:
- List books and download them to the device.
- Manage books (delete from library, add to collections, etc)
- Store tracking metadata for each book (completion, reading location, last read, etc)

I've already implemented a pretty complete implementation relying on the Calibredb commandline tool, but it's a little fragile and has been hard to work with (CalibreDb messes up the directory structure if different OSes interact with the library :-/ ).
I'm also storing books on Backblaze, with only the metadata.db file synced to the server where CalibreWeb is running. 
I ended up forking this project and have started re-implementing it by relying on db.py instead.

There's a few open questions that I'm hoping to get some opinions on:
1. Should I remain in a fork, since I mostly really care about the db.py abstraction. Or should I try to get it included into the main project?

2. Why are shelves/collections managed in a separate database?
Historically, Kobo users have been using Calibre custom columns to manage their devices shelves. A calibre plugins wasdeveloped to keep the device shelves in sync with a given "MyShelves" custom column (see https://wiki.mobileread.com/wiki/Kobo_Shelves_and_Collections).
In my CalibreDb-based implementation, I've also been relying on custom columns to replicate this behavior. This works for the most part, except that the Sync protocol expects shelves to have unique ids. I've faked this by simply using the shelf name as the id instead of a uuid string, and this has been working so far.

I'm wondering if I should try to replicate that shelf management style here, or if I should rely on the secondary database that CalibreWeb seems to be using to manage collections.
I'm more inclined towards the first since that what every Kobo Calibre user is supposedly doing now, however it wouldn't match the behavior of the CalibreWeb web interface.

**Comment by shavitmichael** [comment]
On second thought, the exact use-case I'm seeking may be difficult with Calibre-Web.

I'm currently managing my Library using the desktop Calibre app, and syncing the metadata.db file to a separate server running the Kobo protocol. If I were reading the metadata.db file using Calibre-Web however, I imagine it wouldn't catch that the metadata.db file has been changed from underneath it; is that correct?

**Comment by OzzieIsaacs** [comment]
Sorry for the long time before replying.
I tried to take a look how your code works and what it does, but unfortunately you didn't commit it to github (as far as I could see). So I don't know how much effort it would take to make it "calibre-web"-ish and how big the impact to calibre-web would be. I personally use a kobo reader and I would like to see an easier way of syncing my ebooks to my Kobo.
Regarding Backblaze: A definite no for integrating this, I still regred the gdrive integration ( I personally don't use it. The original code was difficult to setup, I tried to improve it, now it's better, but I break it usually while tring to improve it. And I'm having always users having trouble to set it up, or to use it. Same with LDAP...).
2. Easy answer: It was like that as I started coding for this project. You could ask the guy who started this project 6 years ago... I think it might be because kindle can't handle the shelfs, or they are not support 6 years ago. I think one of the advantages of the "calibre-web shelfs" is that it'S multiuser capable (I think the calibre version can't handle this?). 
Calibre-web looks easier (hopefully) from the users perspective, but in the background there a many options and supported usecases (multiuser, linked custom column for read/unread status, access controll, opds feed, integrated reader), so we shouldn't focus just on the "kobo-universe" but also try to not forget the other readers.
3. Regarding changed metadata from underneath. This should normally work, you can access the metadata.db from calibre while having it open in calibre-web and you get the changes from calibre with a press of the reload button (but not the other way, because calibre locks the database). But it's kind of dangerous, if both sides change things at the same time (this could be prevented maybe by a mechanism similar to the public reconnect point (#934).

So I'd love to see your code, also possible per e-mail if you don't want to make it open to the public (yet).

**Comment by shavitmichael** [comment]
Thanks a lot for the input!

The calibredb based implementation is definitely too ugly to make public....
However, I had already started on a small calibre-web based prototype when I opened this issue, and I can definitely share that since it's a lot cleaner: https://github.com/shavitmichael/calibre-web/blob/kobo/cps/kobo.py.

There's still a few artifacts from the previous implementation there that I need to clean-up:
1. I've been relying on custom columns to store reading state, but as per your suggestion I should probably use the Calibre-web reading bits instead :).
2. The set-up I've been using for the past few months has been relying on backblaze to act as a cdn for books and covers (the metadata.db file on the other hand is read normally by the server). The prototype I pushed assumes that set-up, however I can probably have it work in a non-backblaze set-up with a little more effort. (And move out the backblaze portion to a fork or hidden option).

> 3. Regarding changed metadata from underneath. This should normally work, you can access the metadata.db from calibre while having it open in calibre-web and you get the changes from calibre with a press of the reload button (but not the other way, because calibre locks the database). But it's kind of dangerous, if both sides change things at the same time (this could be prevented maybe by a mechanism similar to the public reconnect point (#934).

That's perfect! As long as reloading doesn't close the connection, I could trigger it at the beginning of the Kobo sync protocol.
